glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+    base_identifier: &str,
+    label: DynamicLabelType,
+) -> String {
+    // Pick out the key and category from the supplied label
+    if let Some((mut key, mut category)) = separate_label_into_key_and_category(&label) {
+        // Loop through the stores we expect to find this metric in, and if we
+        // find it then just return the full metric identifier that was found
+        for store in &meta.inner.send_in_pings {
+            if glean.storage().has_metric(meta.inner.lifetime, store, key) {
+                return combine_base_identifier_and_labels(base_identifier, key, category);
+            }
+        }
+
+        // Count the number of distinct keys and categories already recorded, we can figure out which
+        // one(s) to check based on the label variant.
+        let (key_count, category_count) = count_keys_and_categories(meta, glean);
+        match label {
+            DynamicLabelType::Label(ref label) => {
+                record_error(
+                    glean,
+                    meta,
+                    ErrorType::InvalidLabel,
+                    format!("Invalid `DualLabeledCounter` label format: {label:?}"),
+                    None,
+                );
+                key = OTHER_LABEL;
+                category = OTHER_LABEL;
+            }
+            DynamicLabelType::KeyOnly(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::CategoryOnly(_) => {
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::KeyAndCategory(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+        }
+        combine_base_identifier_and_labels(base_identifier, key, category)
+    } else {
+        record_error(
+            glean,
+            meta,
+            ErrorType::InvalidLabel,
+            "Invalid `DualLabeledCounter` label format, unable to determine key and/or category",
+            None,
+        );
+        combine_base_identifier_and_labels(base_identifier, OTHER_LABEL, OTHER_LABEL)
+    }
+}
+
+fn label_is_valid(label: &str, glean: &Glean, meta: &CommonMetricDataInternal) -> bool {
+    if label.len() > MAX_LABEL_LENGTH {
+        let msg = format!(
+            "label length {} exceeds maximum of {}",
+            label.len(),
+            MAX_LABEL_LENGTH
+        );
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else if label.chars().any(|c| !c.is_ascii() || c.is_ascii_control()) {
```

Comment:
As mentioned before we should not need the ascii check here.
Though we should check for `\x1e` and reject it.